txstratum/healthcheck.py

```diff
@@ -0,0 +1,265 @@
+            health_check.output = "couldn't connect to fullnode"
+
+            return health_check
+
+
+class MiningHealthCheck(ComponentHealthCheckInterface):
+    """
+    This class receives a manager instance as parameter and implements a health check method for it.
+
+    It will:
+    - Check that the manager has at least 1 miner
+    - Check that all 'tx_jobs' in the manager in the last 5 minutes have a status of done
+      and have a total_time lesser than 10 seconds.
+    - Check that all miners have submitted at least 1 job in the last 1 hour.
+      Note that subimitting a job doesn't mean solving a tx.
+
+    If at least one of the 'tx_jobs' has a 'total_time' of more than 10 seconds, the health check
+    will be returned as 'warn'.
+
+    If some of the 'tx_jobs' has status different than 'done', the health check will be returned
+    as 'fail'.
+
+    If there are no miners, the health check will be returned as 'fail'
+
+    If there are miners, but any of them has submitted a job in the last 1 hour, the health check
+    will be returned as 'fail'
+    """
+
+    def __init__(self, manager: "TxMiningManager") -> None:
+        """Init the class with the manager instance."""
+        self.manager = manager
+        self.last_manager_status = ComponentHealthCheck(
+            component_name="manager",
+            component_type=ComponentType.INTERNAL,
+            status=HealthCheckStatus.PASS,
+            output="everything is ok",
+        )
+
+    async def get_health_check(self) -> ComponentHealthCheck:
+        """Return the manager health check status."""
+        # Check that the manager has at least 1 miner
+        if not self.manager.has_any_miner():
+            return ComponentHealthCheck(
+                component_name="manager",
+                component_type=ComponentType.INTERNAL,
+                status=HealthCheckStatus.FAIL,
+                output="no miners connected",
+            )
+
+        # Check that all the miners submitted in the last 1 hour. If not, return failed.
+        if not self.manager.has_any_submitted_job_in_period(period=3600):
+            return ComponentHealthCheck(
+                component_name="manager",
+                component_type=ComponentType.INTERNAL,
+                status=HealthCheckStatus.FAIL,
+                output="no miners submitted a job in the last 1 hour",
+            )
+
+        if not self.manager.tx_jobs:
+            # We just return the last saved status in case there are no jobs in the last 5 minutes
+            return self.last_manager_status
+
+        for job in self.manager.tx_jobs.values():
+            if job.is_failed():
+                self.last_manager_status.update(
+                    {
+                        "status": HealthCheckStatus.FAIL,
+                        "output": "some tx_jobs in the last 5 minutes have failed",
+                    }
+                )
+
+                return self.last_manager_status
```

Comment:
If I'm not mistaken, if we have a failed job and then stay, say, 10 minutes without any new jobs, we would respond with the `"some tx_jobs in the last 5 minutes have failed"` message even though there were no jobs in the last 5 minutes, which could be misleading during debugging.

Not sure if just changing this message is enough. It would probably better to create a new status for the "no jobs in the last 5 minutes" case, adding a specific output message.

Reply:
I've fixed this in https://github.com/HathorNetwork/tx-mining-service/pull/104/commits/379470f4c215148a0b13fec8699af290a719a3ed